### ISSUE TITLE
Watch chart type attribute?

### ISSUE THREAD

**Issue by Rodeoclash** [paradox41/angular-chartist.js#25]
What do you think about watching the chart type and rebuilding the chart if it changes?

**Comment by Rodeoclash** [comment]
Here you go!

http://plnkr.co/edit/d1Ki3wqScZtYHV9i5tWC?p=preview

If you change the dropdown from Bar > Line then it should update the chart (by watching the value).

As it's just a string, you could use an attr observe? I'm not sure if that's any less resource intensive then a watch though.

**Comment by paradox41** [comment]
hmm

Unfortunately, I think the only fix will have to be a `$watch` as even bidirectional binding wouldn't force the chart to redraw.

The question now is: Do we simply `$watch` __all__ the attributes for changes and redraw the chart or just `$watch` the two mandatory attributes - I already have a `$watch` on data - and leave the other attributes untouched.

I just wonder which scenario better covers most common use cases...

**Comment by Rodeoclash** [comment]
Well I would prefer one watch for all attributes as this would create less watchers over all. You're using a deep watch on the data attributes anyway so it's trivial to just return an array of all the attributes.

The other consideration is if it's cheaper to update the Chartist chart with new data when it changes, vs. tearing the chart down and rebuilding it. You're going to have to tear the chart down and rebuild when the type changes but I'm not sure if Chartist has an update function for new data, I think from memory that it does.

If it's the case, you might want to have to watch functions, one for the data changing (which calls the update) and one for the chart type (which totally rebuilds the chart).

I can have a crack at this if you want and send you a pull request with my implementation.

**Comment by paradox41** [comment]
Yeah please do. 

I never updated that watch when Chartist added an update call, so that would be great.

**Comment by Rodeoclash** [comment]
https://github.com/Crosstivity/ng-chartist.js/blob/master/dist/ng-chartist.js

I'm not going to submit it has a pull request as it's a terrible hack, I also based it off an older version of your code as I ran into some issues using the charts inside an ng-if statement with the newer version.

I'll try and come back to this a bit later but we needed something available immediately for demos.

I also changed some of the attribute names to make them more consistent. Also, what was the reason why you went with ampersands in the scope bindings instead of =? I've never seen that before. Is it a performance thing?

**Comment by paradox41** [comment]
Alright, not a problem, I can probably go ahead and make the fix myself.

Out of curiosity, what about the attribute names did you find inconsistent?

The reason we went with `&` instead of `=` is because Chartist will actually transform whatever data / options it is passed, and we don't want those changes make it back to the user's code. Having bi-directional binding would cause data in your controllers to change.

**Comment by Rodeoclash** [comment]
I removed the "chartist" prefix from some of them and prefixed them all with "chart" (e.g. chartistChartOptions -> chartOptions, chartistResponsiveOptions -> chartResponsiveOptions).

Ah, that's interesting about the ```&```. I thought it would still return the same object by reference (isn't that how the watch is working at the moment? I'm still altering the original object that's been passed in and the watch is firing because of the change to it?)

I figured you'd have to use ```angular.copy()``` to get a true copy of the object and not a reference to it. Good to know though!

**Comment by paradox41** [comment]
So the chartist prefix was there to act as a namespace for this directive.
I think that something generic like chart-options could potentially cause
problems, but the risk is low. Namespacing is a good idea either way for
something like this.

The rest of the options were so named in order to match chartist's API -
http://gionkunz.github.io/chartist-js/api-documentation.html.

I just realized that chart-options isn't really a thing, it is exposed
simply as options. I should probably change that to chartistOptions to be
consistent.

**Comment by paradox41** [comment]
Btw, I'll go ahead and implement a fix for this hopefully over the weekend.

**Comment by Rodeoclash** [comment]
Excellent! Thanks for your work on it btw, was very straight forward to get things working (and we had tried a lot of other charting libraries /w angular wrappers around them before settling on this).

**Comment by paradox41** [comment]
:smile: Thanks for the compliment, I appreciate it. And thanks for continuing to provide feedback.

**Comment by paradox41** [comment]
Haven't forgotten about this btw, just didn't get a chance this last weekend.

**Comment by Rodeoclash** [comment]
No problem. I'm a bit less busy at work now so maybe in the next couple of days I could do a proper pull request on this if you don't get time.

**Comment by paradox41** [comment]
http://plnkr.co/edit/KLz0i2Bef055sxbPnQtw?p=preview

The code is now on master. If you don't mind, please check it and let me know if it works for you.

If it's all good, I will release it.

**Comment by Rodeoclash** [comment]
Works great!

**Comment by Rodeoclash** [comment]
Actually, sorry I had a cached version. I'm a bit confused about your implementation still, e.g:

```scope.$watchGroup(['data', 'chartType', 'chartOptions'], function (newDataArray, oldDataArray) {```

The ```data``` on the scope in this watch appears to be a function that when executed, returns the scope data from the parent. Does watching this automatically execute the function or are you just watching the method itself? Either way, I'm having trouble getting it working correctly.

Thanks for adding this functionality but I think I'm going to stick with our own implementation, mainly as we don't have a concern with Chartist altering the data when processing it and removing the parent scope calling functionality and simply relying on the 2 way binding greatly simplifies the directive.

**Comment by paradox41** [comment]
Hmm. Let me look into that.

**Comment by paradox41** [comment]
So apparently `$watchGroup` does not allow for deep watching (third argument for `$watch`)...a little wtf if I'm honest, means I am gonna have to use two watches...but I guess I don't have a choice...

**Comment by Rodeoclash** [comment]
Could you use a regular $watch and just return array of the items via a function?

**Comment by paradox41** [comment]
Mind sharing a code snippet? Not sure how I feel about that to be honest.

**Comment by Rodeoclash** [comment]
Something like:

```
scope.$watch('[data, type, options, events]', function () {
  ...
}, true);
```

**Comment by Rodeoclash** [comment]
Hmm, I just noticed this commit as well:

https://github.com/gionkunz/chartist-js/issues/106

Maybe you don't need to have the detach method before updating the chart.

**Comment by paradox41** [comment]
So, TIL that actually works...also a little scary if I'm honest, but it makes sense. Fairly elegant solution to be honest. Thanks!

**Comment by Rodeoclash** [comment]
Oh I see, detach is just for the events, just ignore me :)

**Comment by paradox41** [comment]
So, playing with it a little more, I realize it might not work as elegantly as we hope, perhaps you can help me out.

Notice [here](https://github.com/paradox41/angular-chartist.js/blob/334420bf3706ee6d65ceabd0e2df7a4f299b2269/src/angular-chartist.js#L48) that `scope.data` is __not__ in quotes and I have a feeling that there is some angular magic going on to get that to work. If you put it in quotes, it stops working. 

Perhaps I am implementing it incorrectly, but if you wanna take a look, I'd appreciate it.

For now, I am gonna fix the bug related to data not updating correctly but the leave the implementation split out into two `$watches`

**Comment by Rodeoclash** [comment]
Here's my very quick rewrite of the directive:

```
angular.module('crosstivityApp').directive('chartist', function (
	
) {
	return {
		restrict: 'EA',
		scope: {
			data: "=",
			type: "@",
			options: "=",
			events: "="
		},
		link: function (
			scope:Crosstivity.IChartistScope, 
			element:ng.IAugmentedJQuery,
			attrs:ng.IAttributes
		) {

			var chart,
				detach = function () {
					if (chart) {
						chart.detach();
					}
				},
				bindEvents = function () {
					Object.keys(scope.events).forEach(function (eventName) {
						chart.on(eventName, scope.events[eventName]);
					});
				};

			scope.$watch('[data, type, options, events]', function () {
				detach();
				chart = window['Chartist'][scope.type](element[0], scope.data, scope.options);
				bindEvents();
			}, true);

			scope.$on('$destroy', function () {
				detach();
			});

		}
	}
});
```

I think the way your calling the ```&``` functions to copy the data is adding unnecessary complexity to the directive, I'm not even sure if it will actually copy the data and I suspect it's still being passed by reference anyway.

If you don't want Chartist to alter the data structure, I'd look at calling ```angular.copy``` (https://docs.angularjs.org/api/ng/function/angular.copy) inside the watch function then passing the output of that to Chartist to render.

Also, I added a scope.$destroy because I suspect the events were not getting cleaned up correctly when the scope was being destroyed.

Mines a terrible hack though, I whipped it up in about 20 minutes but it seems to work cleanly (although I suspect, it's not that efficient as I'm not using the ```.update()``` function for Chartist.

**Comment by paradox41** [comment]
So, the choice to go with the `&` is, in my opinion, the correct way to expose this interface for a couple of reasons:

- There is no bi-directional binding taking place here; this directive is read-only. Regardless of whether or not Chartist actually modifies the data, it is more idiomatic to use the `&` as the consumer of this directive can be sure that no changes made to the data in the directive will propagate back up to the parent scope.
- Setting up bi-directional binding doesn't remove the need for a `$watch` - which in my mind would the be the only benefit - the data still needs to be explicitly watched so that the chart can be updated.

Also, the reason we might not be seeing any changes in data (with bi-directional binding) is because Chartist itself is keeping a reference to the original data somewhere and only modifying a deep copy that __it__ creates. 

I should definitely add a `$destroy` though to clean up, that is something that does need to be added.

**Comment by Rodeoclash** [comment]
I had to write a Plunkr to try and figure this out for myself. Calling ```&``` will not copy the object being passed in, it's still passed by object reference.

Here's a Plunkr where a directive fetches the data object of its parent controller by calling ```&```. You'll see that modifications to the data object in the directive still affect the parent controller.

http://plnkr.co/edit/nyxYRTb3CIxItGXjIcE1?p=preview

It seems to me that your watch function is simply watching the dereferenced contents of the ```&``` which happens to be the parent controller / directives data. The reason you had the problem watching all the scope items in an array is correct, you're watching the string contents of a function which returns data from the parent scope. However, even if the watch function executed the ```&``` function though, you'd still get an object that was referenced to the parent controllers data.

**I totally agree with making the directive read only** but I don't think ```&``` is achieving that. I think if you want to do that, you'll need to explicitly copy the object using ```angular.copy```. You could do something like:

```
scope.$watch(scope.data, function (newData, oldData) {
  if (newData !== oldData) {
    chart.update(angular.copy(newData));
  }
}, true);
```

I do feel like I could be totally missing something here but I'm 99% sure of the way that ```&``` works. It would be great if you could improve my understanding of it a bit if I'm wrong :)

**Comment by paradox41** [comment]
So I've always thought that the API for directives is really strange. I've actually experienced what you are seeing there very early on and wondered what, if anything, distinguishes `=` from `&`. 

I think there are two separate conversations taking place here:
1. What exactly does `&` do?
2. Should we use `=` in the directive instead of `&`?

Regarding number one, looking back at this thread, I guess there was a misunderstanding somewhere as I am not sure the notion of `&` copying the data came from. I'm not sure if I was the one that suggested that, but if I was I must have had a moment of insanity. Per the docs, `&` allows you to "evaluate an expression in the context of the parent scope." Something like a function call normally calls to mind, but an expression can be anything really.

Regarding number two, I think the answer to this is still no. Along with the reasons I listed above I should also point out that I do not think it is the responsibility of the directive to clone the data so that no changes make it back up to the parent scope. That is the responsibility of Chartist. Think of it this way, without Angular, would people have to deep clone their own data and then pass that to Chartist? 

As I mentioned above, Chartist is probably doing this already since we are not actually seeing any changes to the data being propagated up (and I am using `&` and not deep copying). 

As far as the API this directive __should__ expose, `&` is the correct choice as there is no bi-directional binding. Exposing an `=` would be disingenuous as it would lead the consumer of this directive to believe that there are transformations taking place in the directive that would be visible in the parent controller. 

Your plunkr demonstrates that `&` does not prevent changes from being propagated, but the `&` does not carry with it the assumption that they __will__ be propagated, and I think that is the important distinction.

**Comment by Rodeoclash** [comment]
Really the difference between ```=``` and ```&``` is fairly small. ```&``` is just syntactic sugar around the ```$parse``` service. I've only ever seen it used with event related code as it makes it easy to return arguments to the parents scope function.

It seems like this is a philosophical discussion now if Chartist doesn't modify the referenced object (for some reason I thought it did, which is what has caused this entire thread).

I still think using ```=``` is the correct approach as it saves needing to (dereference? That's a Clojure term for kind of what his is doing) the variable by calling it as well as making it play nice with your ```$watch```.

Back to the original question of how to handle everything in one watch, you should be able to do something like (I'm writing a bit from memory here now)

```
scope.$watch(function () {
  return [Ctrl.data, scope.chartType];
}, function (newArr, oldArr) {
  ...
}, true);
```

That will deep watch the data and the chartType in one ```$watch```. I think this ticket is pretty much done right? :)

**Comment by paradox41** [comment]
Nice. I will attempt to clean up the code a bit more (based on your snippet there) and then release as 3.0.

FWIW, this was a good discussion imo and I am glad we had it :)

**Comment by Rodeoclash** [comment]
Thanks for the code! We're making good use of it on our dashboard!

![screen shot 2015-01-21 at 11 58 13 am](https://cloud.githubusercontent.com/assets/63807/5829336/e20f8e20-a164-11e4-8960-4a9221c5b7a8.png)
